File: java/src/main/java/com/urlshortener/core/shorturl/impl/ShortUrlServiceImpl.java
package com.urlshortener.core.shorturl.impl;

import com.urlshortener.core.dtos.CreateShortUrlDto;
import com.urlshortener.core.dtos.ShortUrlDto;
import com.urlshortener.core.dtos.UpdateShortUrlDto;
import com.urlshortener.core.exceptions.EntityNotFoundException;
import com.urlshortener.core.namegenerator.NameGenerator;
import com.urlshortener.core.shorturl.ShortUrlService;
import com.urlshortener.dal.entities.Constraints;
import com.urlshortener.dal.entities.ShortUrl;
import com.urlshortener.dal.repositories.shorturl.ShortUrlRepository;
import com.urlshortener.dal.repositories.user.UserRepository;
import lombok.NonNull;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import javax.enterprise.context.RequestScoped;
import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.transaction.UserTransaction;
import java.util.List;
import java.util.stream.Collectors;

import static com.urlshortener.core.util.TransactionUtils.withUniqueConstraintHandling;

@RequestScoped
public class ShortUrlServiceImpl implements ShortUrlService {

    final ShortUrlRepository shortUrlRepository;

    final UserRepository userRepository;

    final NameGenerator nameGenerator;

    final UserTransaction transaction;

    @ConfigProperty(name = "urlshortener.generatedNameLength")
    int generatedNameLength;

    @Inject
    public ShortUrlServiceImpl(ShortUrlRepository shortUrlRepository, UserRepository userRepository, NameGenerator nameGenerator, UserTransaction transaction) {
        this.shortUrlRepository = shortUrlRepository;
        this.userRepository = userRepository;
        this.nameGenerator = nameGenerator;
        this.transaction = transaction;
    }

    @Override
    public ShortUrlDto findByShortname(@NonNull String shortName) {
        var shortUrl = this.shortUrlRepository.findByShortName(shortName);
        if (shortUrl == null) {
            throw new EntityNotFoundException(notFoundByShortNameMessage(shortName));
        }

        return toDto(shortUrl);
    }

    @Override
    public List<ShortUrlDto> findAll(long userId) {
        var shortUrls = this.shortUrlRepository.findAllForUser(userId);
        return shortUrls.stream().map(ShortUrlServiceImpl::toDto).collect(Collectors.toList());
    }

    @Override
    public ShortUrlDto findById(long id, long userId) {
        return toDto(this.getById(id, userId));
    }

    @Override
    public ShortUrlDto update(@NonNull UpdateShortUrlDto updateShortUrlDto) {
        return withUniqueConstraintHandling(this.transaction, Constraints.shortUrlConstraint,
                                            shortUrlAlreadyExistsMessage(updateShortUrlDto.getShortName()), () -> {
                    var persistedEntity = this.getById(updateShortUrlDto.getId(), updateShortUrlDto.getUserId());

                    if (updateShortUrlDto.getShortName() != null) {
                        persistedEntity.setShortName(updateShortUrlDto.getShortName());
                    }

                    if (updateShortUrlDto.getUrl() != null) {
                        persistedEntity.setUrl(updateShortUrlDto.getUrl());
                    }

                    return toDto(persistedEntity);
                });
    }

    @Transactional
    @Override
    public void delete(long id, long userId) {
        this.shortUrlRepository.deleteById(id, userId);
    }

    @Override
    public ShortUrlDto create(@NonNull CreateShortUrlDto createShortUrlDto) {
        return withUniqueConstraintHandling(this.transaction, Constraints.shortUrlConstraint,
                                            shortUrlAlreadyExistsMessage(createShortUrlDto.getShortName()), () -> {
                    var shortName = createShortUrlDto.getShortName() != null ? createShortUrlDto.getShortName() : this.getUniqueShortName();
                    var user = this.userRepository.findById(createShortUrlDto.getUserId());

                    if (user == null) {
                        throw new EntityNotFoundException(userNotFoundByIdMessage(createShortUrlDto.getUserId()));
                    }

                    return toDto(this.shortUrlRepository.merge(
                            new ShortUrl(null, createShortUrlDto.getUrl(), shortName, user)));

                });
    }

    private ShortUrl getById(long id, long userId) {
        var shortUrl = this.shortUrlRepository.findByIdAndUserId(id, userId);
        if (shortUrl == null) {
            throw new EntityNotFoundException(notFoundByIdMessage(id));
        }

        return shortUrl;
    }

    private static ShortUrlDto toDto(ShortUrl shortUrl) {
        return new ShortUrlDto(shortUrl.getShortName(), shortUrl.getUrl(), shortUrl.getId(),
                               shortUrl.getUser().getId());
    }

    private String getUniqueShortName() {
        String generatedName;

        do {
            generatedName = this.nameGenerator.generateName(this.generatedNameLength);
        } while (this.shortUrlRepository.countByShortName(generatedName) != 0L);

        return generatedName;
    }

    private static String notFoundByIdMessage(long id) {
        return "Short url with id " + id + " not found";
    }

    private static String notFoundByShortNameMessage(String shortName) {
        return "Short url with name " + shortName + " not found";
    }

    private static String shortUrlAlreadyExistsMessage(String shortName) {
        return "Short url with name " + shortName + " already exists";
    }

    private static String userNotFoundByIdMessage(long id) {
        return "user with id " + id + " not found";
    }
}

File: src/main/kotlin/com/urlshortener/core/services/shorturl/impl/ShortUrlServiceImpl.kt
package com.urlshortener.core.services.shorturl.impl

import com.urlshortener.core.dtos.CreateShortUrlDto
import com.urlshortener.core.dtos.ShortUrlDto
import com.urlshortener.core.dtos.UpdateShortUrlDto
import com.urlshortener.core.exceptions.EntityModificationException
import com.urlshortener.core.exceptions.EntityNotFoundException
import com.urlshortener.core.exceptions.ShortNameAlreadyExistsException
import com.urlshortener.core.services.namegenerator.NameGenerator
import com.urlshortener.core.services.shorturl.ShortUrlService
import com.urlshortener.core.util.execute
import com.urlshortener.core.util.getException
import com.urlshortener.dal.entities.ShortUrl
import com.urlshortener.dal.entities.shortNameConstraint
import com.urlshortener.dal.repositories.ShortUrlRepository
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.sql.SQLException
import javax.enterprise.context.RequestScoped
import javax.inject.Inject
import javax.transaction.UserTransaction

@RequestScoped
class ShortUrlServiceImpl(
    @Inject private val shortUrlRepository: ShortUrlRepository,
    @Inject private val nameGenerator: NameGenerator,
    @Inject private val transaction: UserTransaction,
) : ShortUrlService {

    @ConfigProperty(name = "urlshortener.generatedNameLength")
    private var generatedNameLength: Int = 0

    override fun findByShortname(shortName: String): ShortUrlDto =
        shortUrlRepository.findByShortName(shortName)?.toDto() ?: throw EntityNotFoundException(
            notFoundByShortNameMessage(shortName)
        )

    override fun findAll(userId: String): List<ShortUrlDto> =
        shortUrlRepository.findAllByUserId(userId).map { it.toDto() }

    override fun findById(id: Long, userId: String) = getById(id, userId).toDto()

    override fun update(entity: UpdateShortUrlDto): ShortUrlDto =
        withShortNameConstraintHandler(entity.shortName, "update") {
            val persistedEntity = getById(entity.id, entity.userId)

            entity.shortName?.let {
                persistedEntity.shortName = it
            }

            entity.url?.let {
                persistedEntity.url = it
            }

            persistedEntity.toDto()
        }

    override fun create(createShortUrlDto: CreateShortUrlDto): ShortUrlDto =
        withShortNameConstraintHandler(createShortUrlDto.shortName, "create") {
            shortUrlRepository.merge(
                ShortUrl(
                    createShortUrlDto.url,
                    createShortUrlDto.shortName ?: getUniqueShortName(),
                    createShortUrlDto.userId
                )
            ).toDto()
        }

    override fun delete(id: Long, userId: String) =
        if (shortUrlRepository.existsByIdAndUserId(id, userId)) {
            shortUrlRepository.deleteById(id)
        } else throw EntityNotFoundException(notFoundByIdMessage(id))

    private inline fun withShortNameConstraintHandler(
        shortName: String?,
        modificationType: String,
        block: () -> ShortUrlDto
    ): ShortUrlDto = try {
        transaction.execute(block)
    } catch (ex: Exception) {
        if (shortName != null) {
            val sqlException = ex.getException<SQLException>()
            if (sqlException != null && sqlException.message?.contains(shortNameConstraint) == true) {
                throw ShortNameAlreadyExistsException(shortName)
            }
        }

        throw EntityModificationException("could not $modificationType short url")
    }

    private fun getById(id: Long, userId: String): ShortUrl =
        shortUrlRepository.findByIdAndUserId(id, userId) ?: throw EntityNotFoundException(
            notFoundByIdMessage(
                id
            )
        )

    private fun ShortUrl.toDto(): ShortUrlDto = ShortUrlDto(shortName, url, id!!, userId)

    private fun getUniqueShortName(): String {
        var generatedName: String

        do {
            generatedName = nameGenerator.generateName(generatedNameLength)
        } while (shortUrlRepository.countByShortName(generatedName) != 0L)

        return generatedName
    }

    companion object {
        private fun notFoundByIdMessage(id: Long) = "Short url with id $id not found"
        private fun notFoundByShortNameMessage(shortName: String) = "Short url with name $shortName not found"
    }
}